Cpp/cbg-executor_ping-pong/README.md

```diff
@@ -0,0 +1,58 @@
+
+![](doc/ping_pong_diagram.png)
+
+The Ping node can be configured to send messages at a configured rate. The Pong node takes these ping messages and replies each of them. Before sending the reply, it can be configured to burn cycles (thereby varying the processor load) to simulate some message processing. We provide bash scripts to test intra-process and inter-process communication scenarios, wherein the nodes are co-located either in one process or two processes, respectively. These scripts also vary the message rates and processor loads. After each run, the Ping node outputs the measured the throughput of real-time and best effort messages.
+
+
+## Running the test bench
+
+After building the test bench with [colcon](https://github.com/ros2/ros2/wiki/Colcon-Tutorial), the Ping and Pong nodes may be either started in one process or in two processes. Please note that the test bench requires sudo privileges to be able to change the thread priorities using `pthread_setschedparam(..)`. Start the executable by
+
+```
+ros2 run cbg-executor_ping-pong_cpp ping-pong args...
+```
+
+where `args...` are five or six arguments as follows:
+
+```
+[type] [rt_ping_period_us] [be_ping_period_us] [rt_busyloop_us] [be_busyloop_us] [cpu_id]
+
+type: determines the nodes included in this process:
+  i: ping node only
+  o: pong node only
+  io: ping node and pong node
+
+rt_ping_period_us: microseconds between publishing of ping messages by real-time thread
+                   in ping node
+be_ping_period_us: microseconds between publishing of ping messages by best-effort
+                   thread in ping node
+rt_busyloop_us:    microseconds of computation by real-time thread in pong node before
+                   answering with pong
+be_busyloop_us:    microseconds of computation by best-effort thread in pong node before
+                   answering with pong
+cpu_id (optional): pins both, real-time thread and best-effort thread, to the given cpu
+```
+
+When using type `i`, a second process with type `o` has to be started simultaneously.
+
+The shell run\* scripts in this folder run various experiments in sequence. For pertinent results, we propose to use the [PREEMPT_RT patch](https://wiki.linuxfoundation.org/realtime/start) for the Linux kernel.
+
+
+## Implementation Details
+
+The algorithms of the Ping node and of the Pong node are factored out into classes [_PingSubNode_](include/PingSubNode.hpp) and [_PongSubNode_](include/PongSubNode.hpp) - configurable with regard to the real-time profile and the topic prefix. Thus, the Ping node contains of two instances of the _PingSubNode_ and the Pong node contains two instances of _PongSubNode_. (And the test bench could be easily extended to more than two ping-pong paths.)
```

Comment:
Either "Ping node contains ~~of~~ two instances" or "Ping node *consists* of two instances"